z_login.php

```diff
@@ -16,14 +30,42 @@
 $form->addField ('name',TEXT,30,REQUIRED)			->setLabel (L_USER_NAME);
 $form->addField ('password',PASSWORD,30,REQUIRED)	->setLabel (L_USER_PASSWORD);
 
-$form->addCondition (MYSQL_STATEMENT,"SELECT user_id FROM z_users WHERE name='{name}' AND password='{password}'",L_INVALID_LOGIN);
+$form->addCondition (USER_FUNCTION, function() use ($form) {
+	$password_matches = false;
+	$stored_name = '';
+	$stored_password = '';
+	$name = $form->Fields['name']->user_value;
+	$pass = $form->Fields['password']->user_value;
+	$stmt = $form->Connection->link->stmt_init();
+	$stmt->prepare(
+			"SELECT name, password
+			FROM z_users
+			WHERE name=?");
+	$stmt->bind_param('s', $name);
+	$stmt->execute();
+	$stmt->bind_result($stored_name, $stored_password_hash);
+	$stmt->fetch();
+	$stmt->close();
+	if (preg_match('/^\$\S{1,6}\$/', $stored_password_hash)) {
+		$password_matches = password_verify($pass, $stored_password_hash);
+		if ($password_matches && password_needs_rehash($stored_password_hash, PASSWORD_DEFAULT)) {
+			update_user_pass($name, $pass, $form->Connection->link);
+		}
+	} else {
+		$password_matches = ($pass === $stored_password_hash) && ($stored_password_hash !== "");
+		if ($password_matches) {
+			update_user_pass($name, $pass, $form->Connection->link);
+		}
+	}
+	return $password_matches;
+}, L_INVALID_LOGIN);
 
 $form
 	->addButton (BACK)
 	->addButton (SUBMIT);
 
 $form
-	->addAction (QUERY_TO_SESSION,"SELECT * FROM z_users WHERE name='{name}' AND password='{password}'",L_SESSION_NAME)
+	->addAction (QUERY_TO_SESSION,"SELECT * FROM z_users WHERE name='{name}'",L_SESSION_NAME)
```

Comment:
@torstenroeder Is querying the user name sufficient here? Is the condition above guaranteed to run before any `QUERY_TO_SESSION`, and `QUERY_TO_SESSION` not being run if the condition returns `false`?

Reply:
Yes, username is unique, and `QUERY_TO_SESSION` is only executed when condition is true.